ClimsoftVer4/ClimsoftVer4/frmLogin.vb

```diff
@@ -20,206 +20,203 @@ Imports ClimsoftVer4.Translations
-        ConfigFile()
-        'frmMainMenu.Show()
-        'Me.Hide()
-        'regDataInit()
-    End Sub
-
-    Private Sub Cancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Cancel.Click
-        Me.Close()
-    End Sub
-
-    Sub ConfigFile()
-        Try
+        Dim builder As New System.Data.Common.DbConnectionStringBuilder()
+        Dim connectionString As String
+        Dim dbChoice As String
+        Dim password As String
+        Dim subKey As String
+        Dim username As String
+
+        username = txtUsername.Text
+        password = txtPassword.Text
+
+        ' Ensure username and password are not empty
+        If String.IsNullOrEmpty(username) Or String.IsNullOrEmpty(password) Then
+            MsgBox("Please enter a username and password")
+            Exit Sub
+        End If
 
-            ' Create an instance of StreamReader to read from a file. 
-            ' Then open the configuration file
-            'MsgBox(Application.StartupPath.Replace("\bin\Debug", "\config.inf"))
-            'Using sr As New System.IO.StreamReader(Application.StartupPath.Replace("\bin\Debug", "\config.inf"))
+        updateRememberedUsername()
 
-            ' Dim line As String
-            Dim connectstr As String
-            Dim caption As String
+        dbChoice = cmbDatabases.SelectedItem
+        If String.IsNullOrEmpty(dbChoice) Then
+            MsgBox("Please select a database from the list, or manage database connections")
+            Exit Sub
+        Else
+            subKey = "db_" & dbChoice
+            connectionString = My.Computer.Registry.GetValue(
+                    "HKEY_LOCAL_MACHINE\Software\Climsoft4", subKey, Nothing)
+            If String.IsNullOrEmpty(connectionString) Then
+                MsgBox("Unable to read connection information. Please select ""Manage database connections"" " &
+                       "to check and amend connection details")
+                Exit Sub
+            End If
+        End If
 
-            ' Read from the configuration file
-            'line = sr.ReadLine()
+        ' Check that the connection string, with username and password is accepted by the
+        ' connection String builder (otherwise warn and stop)
+        Try
+            builder.ConnectionString = connectionString
+            builder("uid") = username
+            builder("pwd") = password
 
-            'MsgBox(Server_db_port(line))
+            ' The connection string has historically been stored in this control
+            ' There are other locations in the software that may access this
+            txtusrpwd.Text = builder.ConnectionString & ";Convert Zero Datetime=True"
+        Catch ex As Exception
+            MsgBox("Login failed: " & ex.Message)
+        End Try
 
-            ' Contruct the connection string using the output the file
+        conn.ConnectionString = txtusrpwd.Text
+        Try
+            conn.Open()
```

Comment:
@isedwards the connection is not being closed?

Reply:
Thank you Patrick - this is fixed in #575 ([here](https://github.com/opencdms/Climsoft/pull/575/files#diff-31e5ad9223abb094c0afcb16d231bcceR139))